### DISCUSSION TITLE
Can I use a sidecar as an init container?

### DISCUSSION THREAD

**Discussion by BryanDollery** [openkruise/kruise#839] [Q&A] [answered]
To rephrase the question: is it ok for my sidecar to exit?

I want to use all the features of the kruise sidecar capability to inject init containers. This will enable the init container without having to burden my engineers with specifying an init container in their deployment yaml files.

So, is there a problem with doing this?

I'm asking because I am having problems with my sidecar and I can't yet tell if it's something I've done or if kruise has a problem with my sidecar exiting early.

TIA

**Comment by FillZpp** [comment, marked as answer]
1. You can define `initContainers` in SidecarSet, so that it will inject the init containers when new pods created. Refer to [this doc](https://openkruise.io/docs/user-manuals/sidecarset#sidecarset-features).
2. SidecarSet does not care of the existing sidecar or init containers in early pods.

(I'm not sure if I got all your questions)

**Reply by BryanDollery** [reply]
Yeah, I am aware of that, but I think that's an init container for the sidecar. I basically don't want a sidecar at all, just an init container, but I want to define it centrally rather than in every service that needs it, hence my use-case for OpenKruise.

**Reply by FillZpp** [reply]
> but I think that's an init container for the sidecar.

Of course you could just define `initContainers` in a SidecarSet with no `containers` if you need no sidecar containers.

**Reply by BryanDollery** [reply]
Brilliant! I'll give it a go, thanks.

**Reply by FillZpp** [reply]
You're welcome. 😃

**Reply by qixiaobo** [reply]
Very useful. So glad to see this.
But how can I control the order of init container?

**Reply by FillZpp** [reply]
@qixiaobo Good question. Are the init containers defined in a signal sidecarset or multiple sidecarsets?

**Reply by qixiaobo** [reply]
> @qixiaobo Good question. Are the init containers defined in a signal sidecarset or multiple sidecarsets?

Maybe other team[even cloud provider] defined admission hooks. 
You may have a look at arms or mse

**Reply by qixiaobo** [reply]
> > @qixiaobo Good question. Are the init containers defined in a signal sidecarset or multiple sidecarsets?
> 
> Maybe other team[even cloud provider] defined admission hooks. You may have a look at arms or mse

Arms url below 
https://help.aliyun.com/document_detail/103108.html

**Reply by FillZpp** [reply]
kube-apiserver will sort all mutating webhooks by their names and calls them with the name order.

**Reply by qixiaobo** [reply]
> kube-apiserver will sort all mutating webhooks by their names and calls them with the name order.

Yeah, for me . K8s make sure for different webhooks in alpha order. So the last order webhook may has chance to reorder all sidecars.
